os/win/dnd.cpp

```diff
@@ -0,0 +1,335 @@
+// LAF OS Library
+// Copyright (C) 2024  Igara Studio S.A.
+//
+// This file is released under the terms of the MIT license.
+// Read LICENSE.txt for more information.
+
+#include "os/window.h"
+#include "os/win/dnd.h"
+#include "os/system.h"
+
+#include "clip/clip.h"
+#include "clip/clip_win.h"
+
+#include <shlobj.h>
+
+DWORD as_dropeffect(const os::DropOperation op)
+{
+  DWORD effect = DROPEFFECT_NONE;
+  if (op & os::DropOperation::Copy)
+    effect |= DROPEFFECT_COPY;
+
+  if (op & os::DropOperation::Move)
+    effect |= DROPEFFECT_MOVE;
+
+  if (op & os::DropOperation::Link)
+    effect |= DROPEFFECT_LINK;
+
+  return effect;
+}
+
+os::DropOperation as_dropoperation(DWORD pdwEffect)
+{
+  int op = 0;
+  if (pdwEffect & DROPEFFECT_COPY)
+    op |= os::DropOperation::Copy;
+
+  if (pdwEffect & DROPEFFECT_MOVE)
+    op |= os::DropOperation::Move;
+
+  if (pdwEffect & DROPEFFECT_LINK)
+    op |= os::DropOperation::Link;
+
+  return static_cast<os::DropOperation>(op);
+}
+
+
+gfx::Point drag_position(HWND hwnd, POINTL& pt)
+{
+  ScreenToClient(hwnd, (LPPOINT) &pt);
+  return gfx::Point(pt.x, pt.y);
+}
+
+namespace os {
+
+// HGLOBAL Locking/Unlocking wrapper
+template <typename T>
+class GLock {
+public:
+  GLock() = delete;
+  GLock(const GLock&) = delete;
+  GLock(HGLOBAL hglobal) : m_hmem(hglobal) {
+    m_data = static_cast<T>(GlobalLock(m_hmem));
+  }
+
+  ~GLock() {
+    GlobalUnlock(m_hmem);
+  }
+
+  operator HGLOBAL() {
+    return m_hmem;
+  }
+
+  operator T () {
+    return m_data;
+  }
+
+  SIZE_T size() {
+    return GlobalSize(m_hmem);
+  }
+
+private:
+  HGLOBAL m_hmem;
+  T m_data;
+};
```

Comment:
I think `T` should be the type without the pointer (just like other pointer-like wrappers/smart-pointers):
```suggestion
  GLock(HGLOBAL hglobal) : m_hmem(hglobal) {
    m_data = static_cast<T*>(GlobalLock(m_hmem));
  }

  ~GLock() {
    GlobalUnlock(m_hmem);
  }

  operator HGLOBAL() {
    return m_hmem;
  }

  operator T*() {
    return m_data;
  }

  SIZE_T size() {
    return GlobalSize(m_hmem);
  }

private:
  HGLOBAL m_hmem;
  T* m_data;
};
```

Reply:
I'm trying to do this, but I can't get the DragQueryFile (didn't test getImage() yet) to work when refactored this way.

Reply:
It does compiles with some modifications, but it throws a runtime error

Reply:
I was able to make it work if:
```c++
  operator T() {
    return m_data;
  }
```
But then it complains when calling read_png from getImage()...

Reply:
I think as there is a `GLock<HDROP>` we cannot change the `GLock` and just leave it as it is (receiving a pointer as `T`).